piptools/utils.py

```diff
@@ -71,13 +71,27 @@ def make_install_requirement(name, version, extras, constraint=False):
         constraint=constraint)
 
 
+def is_url_requirement(ireq, repository=None):
+    """
+    Finds if a requirement is a URL
+    """
+    if not ireq.link or 'pypi.python.org' in ireq.link.url or ireq.link.url.startswith('file'):
```

Comment:
I might be missing something, but `ireq.link` will only have a value when using files uri or vcs urls, right?
If that's the case, why check for the presence of index urls and discard?

Reply:
`ireq.link` can also give the link to download a requirement. For example, for `six==1.10.0`, it can be `https://pypi.python.org/packages/c8/0a/b6723e1bc4c516cb687841499455a8505b44607ab535be01091c0f24f079/six-1.10.0-py2.py3-none-any.whl#md5=3ab558cf5d4f7a72611d59a81a315dc8`

Reply:
And if I remove `or 'pypi.python.org' in ireq.link.url`, I get some tests failures  because of that:
```
E           AssertionError: assert 'six==1.10.0' in '#\n# This file is autogenerated by pip-compile\n# To update, run:\n#\n#    pip-compile --output-file requirements.txt...5be01091c0f24f079/six-1.10.0-py2.py3-none-any.whl#md5=3ab558cf5d4f7a72611d59a81a315dc8\nDry-run, so nothing updated.\n'
E            +  where '#\n# This file is autogenerated by pip-compile\n# To update, run:\n#\n#    pip-compile --output-file requirements.txt...5be01091c0f24f079/six-1.10.0-py2.py3-none-any.whl#md5=3ab558cf5d4f7a72611d59a81a315dc8\nDry-run, so nothing updated.\n' = <Result okay>.output
```

Reply:
Ok, I'll have to do some more tests on this, as `ireq.link` seemed to be always `None` when I was installing packages from PyPi. 
If the Pypi url can be present like you said, then the logic here cannot simply exclude `pypi.python.org`, because it's not the only possible PyPi server. We have `pypi.org`, as well as any private pypi server defined by the user.

Reply:
Aren't those private pypi servers in `index_urls` ? This case is  adressed two lines later